storey/table.py

```diff
@@ -703,3 +708,382 @@ def _get_aggregation_default_value(aggregation):
+
+    def aggregate(self, data, timestamp):
+        # add a new point and aggregate
+        for aggregation_metadata in self.aggregates:
+            if aggregation_metadata.should_aggregate(data):
+                curr_value = aggregation_metadata.value_extractor(data)
+                # for aggr in aggregation_metadata.get_all_raw_aggregates():
+                self.aggregation_buckets[f'{aggregation_metadata.name}'].aggregate(timestamp, curr_value)
+
+    def get_features(self, timestamp):
+        result = {}
+        for aggregation_bucket in self.aggregation_buckets.values():
+            result.update(aggregation_bucket.get_features(timestamp))
+
+        return result
+
+
+class AggregationBuckets:
+    def __init__(self, name, explicit_raw_aggregations, hidden_raw_aggregations, virtual_aggregations,
+                 explicit_windows, base_time, max_value, initial_data=None):
+        self.name = name
+        self._explicit_raw_aggregations = explicit_raw_aggregations
+        self._hidden_raw_aggregations = hidden_raw_aggregations
+        self._all_raw_aggregates = set()
+        self._all_raw_aggregates.update(self._explicit_raw_aggregations)
+        self._all_raw_aggregates.update(self._hidden_raw_aggregations)
+        self._virtual_aggregations = virtual_aggregations
+        self.explicit_windows = explicit_windows
+        self.max_value = max_value
+        self.buckets = []
+        self.should_persist = True
+        self.pending_aggr = {}
+        self.storage_specific_cache = {}
+        self.max_window_millis = self.explicit_windows.max_window_millis
+        self.total_number_of_buckets = self.explicit_windows.total_number_of_buckets
+
+        self._need_to_recalculate_pre_aggregates = False
+        self._last_data_point_timestamp = base_time
+        self._current_aggregate_values = {}
+
+        # If a user specified a max_value we need to recalculated features on every event
+        self._precalculated_aggregations = max_value is None
+
+        self.is_fixed_window = isinstance(self.explicit_windows, FixedWindows)
+        if self.is_fixed_window:
+            self._round_time_func = self.explicit_windows.round_up_time_to_window
+        self.period_millis = explicit_windows.period_millis
+        self._window_start_time = explicit_windows.get_window_start_time_by_time(base_time)
+        if self._precalculated_aggregations:
+            for (window_millis, windows_str) in explicit_windows.windows:
+                for aggr in self._all_raw_aggregates:
+                    self._current_aggregate_values[(aggr, window_millis, windows_str)] = AggregationValue(aggr)
+
+        if initial_data:
+            self.last_bucket_start_time = None
+
+            # Initializing the buckets from the stored data and calculating the initial pre aggregates
+            self.initialize_from_data(initial_data, base_time)
+            self._need_to_recalculate_pre_aggregates = True
+            self.calculate_features(base_time)
```

Comment:
Hm this is strange... Shouldn't the result be used by something?

Reply:
it's to init the pre aggregated values in self._current_aggregate_values

Reply:
Might need some refactoring then.